planning/behavior_path_planner/src/utils/path_safety_checker/safety_check.cpp

```diff
@@ -274,6 +298,147 @@ boost::optional<PoseWithVelocityAndPolygonStamped> getInterpolatedPoseWithVeloci
+  const std::vector<T> & path, const double time_horizon, const F & interpolateFunc)
+{
+  std::vector<T> filtered_path;
+
+  for (const auto & elem : path) {
+    if (elem.time < time_horizon) {
+      filtered_path.push_back(elem);
+    } else {
+      break;
+    }
+  }
+
+  const auto interpolated_opt = interpolateFunc(path, time_horizon);
+  if (interpolated_opt) {
+    filtered_path.push_back(*interpolated_opt);
+  }
+
+  return filtered_path;
+};
+
+std::vector<PoseWithVelocityStamped> filterPredictedPathByTimeHorizon(
+  const std::vector<PoseWithVelocityStamped> & path, const double time_horizon)
+{
+  return filterPredictedPathByTimeHorizon(
+    path, time_horizon, [](const auto & path, const auto & time) {
+      return calcInterpolatedPoseWithVelocity(path, time);
+    });
+}
+
+ExtendedPredictedObject filterObjectPredictedPathByTimeHorizon(
+  const ExtendedPredictedObject & object, const double time_horizon,
+  const bool check_all_predicted_path)
+{
+  auto filtered_object = object;
+  auto filtered_predicted_paths = getPredictedPathFromObj(object, check_all_predicted_path);
+
+  for (auto & predicted_path : filtered_predicted_paths) {
+    // path is vector of polygon
+    const auto filtered_path = filterPredictedPathByTimeHorizon(
+      predicted_path.path, time_horizon, [&object](const auto & poses, double t) {
+        return getInterpolatedPoseWithVelocityAndPolygonStamped(poses, t, object.shape);
+      });
+    predicted_path.path = filtered_path;
+  }
+
+  filtered_object.predicted_paths = filtered_predicted_paths;
+  return filtered_object;
+}
+
+ExtendedPredictedObjects filterObjectPredictedPathByTimeHorizon(
+  const ExtendedPredictedObjects & objects, const double time_horizon,
+  const bool check_all_predicted_path)
+{
+  ExtendedPredictedObjects filtered_objects;
+  filtered_objects.reserve(objects.size());
+
+  for (const auto & object : objects) {
+    filtered_objects.push_back(
+      filterObjectPredictedPathByTimeHorizon(object, time_horizon, check_all_predicted_path));
+  }
+
+  return filtered_objects;
+}
+
+bool checkSafetyWithIntegralPredictedPolygon(
+  const std::vector<PoseWithVelocityStamped> & ego_predicted_path, const VehicleInfo & vehicle_info,
+  const ExtendedPredictedObjects & objects, const bool check_all_predicted_path,
+  const IntegralPredictedPolygonParams & params, CollisionCheckDebugMap & debug_map)
+{
+  const std::vector<PoseWithVelocityStamped> filtered_ego_path = filterPredictedPathByTimeHorizon(
+    ego_predicted_path, params.time_horizon);  // path is vector of pose
+  const std::vector<Polygon2d> extended_ego_polygons =
+    createExtendedPolygonsFromPoseWithVelocityStamped(
+      filtered_ego_path, vehicle_info, params.forward_margin, params.backward_margin,
+      params.lat_margin);
+
+  const ExtendedPredictedObjects filtered_path_objects = filterObjectPredictedPathByTimeHorizon(
+    objects, params.time_horizon, check_all_predicted_path);  // path is vector of polygon
```

Comment:
```suggestion
  const auto filtered_ego_path = filterPredictedPathByTimeHorizon(
    ego_predicted_path, params.time_horizon);  // path is vector of pose

  const auto extended_ego_polygons =
    createExtendedPolygonsFromPoseWithVelocityStamped(
      filtered_ego_path, vehicle_info, params.forward_margin, params.backward_margin,
      params.lat_margin);

  const auto filtered_path_objects = filterObjectPredictedPathByTimeHorizon(
    objects, params.time_horizon, check_all_predicted_path);  // path is vector of polygon
```

Reply:
types related predicted path area confusing, so I prefer to make the type explicit:pray: